### PR TITLE
KMS deletion for AWS

### PR DESCRIPTION
**Pull request by Shanu-Rajmaniar-yb**

### CODE REVIEW

File: crc/aws/kms.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    """
+    The service_name variable specifies the AWS service that this class will interact with.
+    """
+
+    default_region_name = "us-west-2"
+    """
+    The default_region_name variable specifies the default region to be used when interacting with the AWS service.
+    """
+
+    jenkins_user = 'arn:aws:iam::454529406029:user/jenkins-slave'
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:23]: this should come as CLI argument

File: crc/aws/kms.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        exception_tags: Dict[str, List[str]],
+    ) -> None:
+        """
+        Initializes the object with filter tags to be used when searching for kms.
+
+        :param dry_run: A boolean variable that indicates whether the class should operate in dry_run mode or not.
+        In dry_run mode, the class will only list the Resources that match the specified filter and exception tags,
+        but will not perform any operations on them.
+        :param filter_tags: dictionary containing key-value pairs as filter tags
+        :type filter_tags: Dict[str, List[str]]
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py:40]: Done

File: crc/aws/kms.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
+# Copyright (c) Yugabyte, Inc.
+
+import logging
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py:4]: Done

File: crc/aws/kms.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                    if self._should_skip_vpc(keys):
+                        continue
+
+                    key_metadata = client.describe_key(KeyId=keys['KeyId'])
+                    key_state = key_metadata['KeyMetadata']['KeyState']
+                    key_des = key_metadata['KeyMetadata']['Description']
+                    # Check if the key meets the filter conditions
+                    if (
+                        key_state == 'Enabled'
+                        and "Yugabyte Master Key" in key_des
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py]: Done, created new arg to be parsed from CLI that should match key description in JSON policy.

File: crc/aws/kms.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                    ):
+                        policy = client.get_key_policy(KeyId=keys['KeyId'], PolicyName='default')['Policy']
+                        policy_json = json.loads(policy)
+                        for ids in policy_json['Statement']:
+                            user_arn = ids['Principal']['AWS']
+                            if ids['Principal']['AWS'] == self.jenkins_user:
+                                print(f"Key {keys['KeyId']} found with user {ids['Principal']['AWS']}")
+                                active_keys.append(keys['KeyId'])
+                except:
+                    print(f"KEY SKIPPED")
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py]: Done, removed print statements and added logging support everywhere. Also added keyname which is being skipped.

File: crc/aws/kms.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                    skipped_keys.append(keys['KeyId'])
+
+            if not self.dry_run:
+                for cmk_id in keys:
+                    client.schedule_key_deletion(KeyId=cmk_id, PendingWindowInDays=14) # Decide for appropriate window
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py]: Done, set the minimum duration to 7 days as default. User can override this by passing from CLI too.

File: crc/aws/kms.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+        if not self.dry_run:
+            logging.warning(
+                f"number of AWS KMS deleted: {len(self.kms_keys_to_delete)}"
+            )
+            logging.warning(f"List of AWS KMS deleted: {self.kms_keys_to_delete}")
+        else:
+            logging.warning(
+                f"List of AWS KMS (Total: {len(self.kms_keys_to_delete)}) which will be deleted: {self.kms_keys_to_delete}"
+            )
```
[the remark this answers is not in the document]
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py]: Done

File: crc/aws/kms.py (hunk 1 above)
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py]: Done

File: crc/aws/kms.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        # Add deleted IPs to deleted_ips list
+        self.kms_keys_to_delete.extend(list(active_keys))
+
+        if not self.dry_run:
+            logging.warning(
+                f"number of AWS KMS deleted: {len(self.kms_keys_to_delete)}"
+            )
+            logging.warning(f"List of AWS KMS deleted: {self.kms_keys_to_delete}")
+        else:
+            a = 0
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:168]: remove this

File: crc.py [hunk 9]
```diff
@@ -35,6 +34,7 @@
 VMS = "VM"
 IPS = "IP"
 KEYPAIRS = "Keypair"
+# KMS = "KMS"
```
Comment (zlareb1-yb) [inline crc.py:37]: why is this commented

File: crc.py [hunk 10]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
         help="InfluxDB connection details in the form of a dictionary. Example: -i or --influxdb {'url': 'http://localhost:8086', 'org': 'Test', 'bucket': 'CRC', 'resource_suffix': 'test'}",
     )
 
+    # Add Argument for Pending Window
+    parser.add_argument(
+        "-pw",
+        "--pending_window",
```
Comment (zlareb1-yb) [inline crc.py:741]: this name is not making sense. It shd be something like `kms_pending_window`

File: crc.py [hunk 11]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
...
+        default=7,
+        choices=range(7, 31),
+        metavar="PENDING_WINDOW",
+        help="The pending window(days) to schedule KMS deletion after this duration. Must be between 7 and 30 inclusive.",
+    )
+
+    # Add Argument for Key Description
+    parser.add_argument(
+        "-k",
+        "--key_description",
```
Comment (zlareb1-yb) [inline crc.py:752]: it should be `kms_key_description`

File: crc.py [hunk 12]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
...
+        type=str,
+        metavar="KEY_DESCRIPTION",
+        help="The string/name to be present in Key description in the Key JSON policy.",
+    )
+
+    # Add Argument for Jenkins Username
+    parser.add_argument(
+        "-u",
+        "--user",
```
Comment (zlareb1-yb) [inline crc.py:761]: it shd be jenkins_user and we shd not be using "`-u`" type bcz it is not explanatory

File: crc.py [hunk 13]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
...
+        type=int,
+        default=7,
+        choices=range(7, 31),
+        metavar="PENDING_WINDOW",
+        help="The pending window(days) to schedule KMS deletion after this duration. Must be between 7 and 30 inclusive.",
+    )
+
+    # Add Argument for Key Description
+    parser.add_argument(
+        "-k",
```
Comment (zlareb1-yb) [inline crc.py:751]: remove -k

File: crc.py [hunk 14]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
         help="InfluxDB connection details in the form of a dictionary. Example: -i or --influxdb {'url': 'http://localhost:8086', 'org': 'Test', 'bucket': 'CRC', 'resource_suffix': 'test'}",
     )
 
+    # Add Argument for Pending Window
+    parser.add_argument(
+        "-pw",
```
Comment (zlareb1-yb) [inline crc.py:740]: remove -pw

File: crc.py [hunk 15]
```diff
@@ -678,6 +735,35 @@ def get_argparser():
         help="InfluxDB connection details in the form of a dictionary. Example: -i or --influxdb {'url': 'http://localhost:8086', 'org': 'Test', 'bucket': 'CRC', 'resource_suffix': 'test'}",
     )
 
+    # Add Argument for Pending Window
```
Comment (zlareb1-yb) [inline crc.py:738]: add README for these new arguments

File: crc/aws/kms.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+    def _should_skip_kms(self, tags: List[Dict[str, str]]) -> bool:
+        """
+        Check if the kms should be skipped based on the exception tags.
+        :param tags: List of tags associated with the kms
+        :type tags: List[Dict[str,str]]
+        :return: True if the kms should be skipped, False otherwise
+        :rtype: bool
+        """
+        if not self.exception_tags:
+            logging.warning("Tags and not present")
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:76]: *Exception tags are not present

File: crc/aws/kms.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+            return False
+
+        for tag in tags:
+            k = tag["Key"]
+            v = tag["Value"]
+
+            if self.exception_tags:
+                if k in self.exception_tags.keys() and (
+                    not self.exception_tags[k] or v in self.exception_tags[k]
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:85]: what is meaning of not self.exception_tags[k]? in which case this may occur?

File: crc/aws/kms.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+                    policy_json = json.loads(policy)
+                    for ids in policy_json["Statement"]:
+                        user_arn = ids["Principal"]["AWS"]
+                        if user_arn == self.jenkins_user:
+                            logging.info(
+                                f"Key {keys['KeyId']} found with user {user_arn}"
+                            )
+                            active_keys.append(keys["KeyId"])
+            except:
+                logging.warning(f"KEY SKIPPED", keys["KeyId"])
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:147]: is this correct syntax?

File: crc/aws/kms.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+        logging.info(f"total number of active Jenkins keys = {len(active_keys)}")
+
+        if not self.dry_run:
+            for cmk_id in active_keys:
+                client.schedule_key_deletion(
+                    KeyId=cmk_id, PendingWindowInDays=self.pending_window
+                )
+                logging.info(f"CMK - {cmk_id} Deleted from AWS console")
+
+        # Add deleted IPs to deleted_ips list
```
Comment (zlareb1-yb) [inline crc/aws/kms.py:159]: fix this msg

**Review by zlareb1-yb — CHANGES_REQUESTED** [review]

File: crc/aws/kms.py (hunk 8 above)
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py:168]: Done

File: crc.py (hunk 10 above)
Comment (Shanu-Rajmaniar-yb) [reply crc.py:741]: yes changed

File: crc/aws/kms.py (hunk 18 above)
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py:147]: Corrected

File: crc/aws/kms.py (hunk 17 above)
Comment (Shanu-Rajmaniar-yb) [reply crc/aws/kms.py:85]: If tag key is present but value is null/falsy.

File: crc.py (hunk 9 above)
Comment (Shanu-Rajmaniar-yb) [reply crc.py]: Missed it, uncommented

File: crc.py (hunk 15 above)
Comment (Shanu-Rajmaniar-yb) [reply crc.py:745]: Done

File: README.md [hunk 20]
```diff
@@ -137,6 +138,10 @@ python crc.py --cloud <cloud_name> --operation_type <operation_type> --resource
 * `slack_notify_users`: Use this option to tag the user in the Slack notification. It is mandatory to pass `--slack_user_label` with this flag. This option only works for GCP disks deletions.
 * `slack_user_label`: Use this option to specify the label to look up in the GCP disks. This option only works for GCP disks deletions.
 * `influxdb`: Use this option to specify InfluxDB connection details. The argument takes a dictionary value, with keys 'url', 'org', 'bucket', and an optional key 'resource_suffix'. Example usage: -i or --influxdb {'url': 'http://localhost:8086', 'org': 'Test', 'bucket': 'CRC', 'resource_suffix': 'test'}. Only works if specified.
+* `kms_pending_window`: Use this option to specify the number of days before the key actually gets deleted. The number of days
+must be between 7 to 30 inclusive.
+* `kms_key_description`: Use this option to match specfic string in kms key description.
+* `jenkins_user`: Use this option to specify the jenkins slave AWS ARN for which keys will be deleted.
```
Comment (zlareb1-yb) [inline README.md:144]: change this to user instead of jenkins_user

**Review by zlareb1-yb — CHANGES_REQUESTED** [review]

File: README.md (hunk 20 above)
Comment (Shanu-Rajmaniar-yb) [reply README.md]: Done, changed to kms_user

File: README.md [hunk 21]
```diff
@@ -137,6 +138,10 @@ python crc.py --cloud <cloud_name> --operation_type <operation_type> --resource
 * `slack_notify_users`: Use this option to tag the user in the Slack notification. It is mandatory to pass `--slack_user_label` with this flag. This option only works for GCP disks deletions.
 * `slack_user_label`: Use this option to specify the label to look up in the GCP disks. This option only works for GCP disks deletions.
 * `influxdb`: Use this option to specify InfluxDB connection details. The argument takes a dictionary value, with keys 'url', 'org', 'bucket', and an optional key 'resource_suffix'. Example usage: -i or --influxdb {'url': 'http://localhost:8086', 'org': 'Test', 'bucket': 'CRC', 'resource_suffix': 'test'}. Only works if specified.
+* `kms_pending_window`: Use this option to specify the number of days before the key actually gets deleted. The number of days
+must be between 7 to 30 inclusive.
+* `kms_key_description`: Use this option to match specfic string in kms key description.
+* `kms_user`: Use this option to specify the jenkins slave AWS ARN for which keys will be deleted.
```
Comment (zlareb1-yb) [inline README.md:144]: Dont mention Jenkins here

Comment (zlareb1-yb) [reply README.md:144]: Also add one example to delete KMS in examples section

**Pull request merged**